Spurious test failure (SIGSEGV) in rust-analyzer `hir-def` test

This has started somewhat recently but now happens pretty often:
```
     Running unittests src/lib.rs (obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/hir_def-f381e8bd2208ec8b)

running 482 tests
test attrs::docs::tests::docs ... ok
test attrs::tests::crate_attrs ... ok
test expr_store::lower::path::tests::fn_like_path_with_coloncolon ... ok
test expr_store::lower::path::tests::hir_to_ast_crate_path ... ok
test expr_store::lower::path::tests::hir_to_ast_path_super_in_middle ... ok
test expr_store::lower::path::tests::hir_to_ast_plain_path ... ok
test expr_store::lower::path::tests::hir_to_ast_trait_ref ... ok
test expr_store::lower::path::tests::hir_to_ast_super_path ... ok
test expr_store::lower::path::tests::hir_to_ast_self_path ... ok
test expr_store::lower::path::tests::hir_to_ast_type_anchor_path ... ok
test expr_store::lower::path::tests::keywords_in_middle_fail_lowering2 ... ok
test expr_store::lower::path::tests::keywords_in_middle_fail_lowering3 ... ok
test expr_store::lower::path::tests::keywords_in_middle_fail_lowering1 ... ok
test expr_store::scope::tests::broken_inner_item ... ok
test expr_store::scope::tests::ref_patterns_contribute_bindings ... ok
test expr_store::scope::tests::match_guard_if_let ... ok
test expr_store::scope::tests::let_chains_can_reference_previous_lets ... ok
test expr_store::scope::tests::macro_inner_item ... ok
test expr_store::scope::tests::test_lambda_scope ... ok
test expr_store::scope::tests::test_call_scope ... ok
test expr_store::scope::tests::test_bindings_after_at ... ok
test expr_store::scope::tests::test_method_call_scope ... ok
test expr_store::scope::tests::test_match ... ok
test expr_store::scope::tests::test_loop_scope ... ok
test expr_store::scope::tests::test_resolve_local_name_declaration ... ok
test expr_store::scope::tests::test_shadow_variable ... ok
test expr_store::scope::tests::test_resolve_local_name ... ok
test expr_store::tests::body::array_element_cfg ... ok
test expr_store::scope::tests::test_resolve_local_name_shadow ... ok
test expr_store::tests::body::async_fn_weird_param_patterns ... ok
test expr_store::scope::tests::while_let_adds_binding ... ok
error: test failed, to rerun pass `-p hir-def --lib`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/hir_def-f381e8bd2208ec8b '--skip=tests::smoke_test_real_sysroot_cargo' --skip=check_code_formatting -Z unstable-options --format json` (signal: 11, SIGSEGV: invalid memory reference)
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 1:08:32
  local time: Mon May 11 13:59:34 UTC 2026
  network time: Mon, 11 May 2026 13:59:34 GMT
##[error]Process completed with exit code 1.
```
Also see discussion on Zulip:
* [#t-compiler/rust-analyzer > test runner segfault on CI](https://rust-lang.zulipchat.com/#narrow/channel/185405-t-compiler.2Frust-analyzer/topic/test.20runner.20segfault.20on.20CI/with/594311587)
* [#t-infra > Debugging spurious CI segfault](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/Debugging.20spurious.20CI.20segfault/with/594301997)

Example of affected PRs:
https://github.com/rust-lang/rust/pull/156357#issuecomment-4416824252
https://github.com/rust-lang/rust/pull/156451#issuecomment-4421411144

--- a/src/ci/docker/scripts/stage_2_test_set2.sh
+++ b/src/ci/docker/scripts/stage_2_test_set2.sh
@@ -12,8 +12,18 @@ if [ "$PR_CI_JOB" == "1" ]; then
   SKIP_TIDY="--skip tidy"
 fi
 
+# The combination of rust.randomize-layout and LLVM 21 is causing random
+# SIGSEGV in the `hir-def` rust-analyzer tests. Since it seems to be fixed in
+# LLVM 22, let's just disable this for now since it has been disruptive for
+# CI. See https://github.com/rust-lang/rust/issues/156460.
+if [[ "$CI_JOB_NAME" == *"llvm-21"* ]]; then
+  echo "CI_JOB_NAME contains 'llvm-21'; skipping rust-analyzer"
+  SKIP_RUST_ANALYZER="--skip src/tools/rust-analyzer"
+fi
+
 ../x.py --stage 2 test \
   ${SKIP_TIDY:+$SKIP_TIDY} \
+  ${SKIP_RUST_ANALYZER:+$SKIP_RUST_ANALYZER} \
   --skip tests \
   --skip coverage-map \
   --skip coverage-run \

